models/user.go

```diff
@@ -423,25 +427,50 @@ func (u *User) NewGitSig() *git.Signature {
 	}
 }
 
-func hashPassword(passwd, salt string) string {
-	tempPasswd := pbkdf2.Key([]byte(passwd), []byte(salt), 10000, 50, sha256.New)
+func hashPassword(passwd, salt, algo string) string {
+	var tempPasswd []byte
+
+	switch algo {
+	case "bcrypt":
+		tempPasswd, _ = bcrypt.GenerateFromPassword([]byte(passwd), bcrypt.DefaultCost)
+		return string(tempPasswd)
+	case "scrypt":
+		tempPasswd, _ = scrypt.Key([]byte(passwd), []byte(salt), 65536, 16, 2, 50)
+	case "argon2":
+		tempPasswd = argon2.IDKey([]byte(passwd), []byte(salt), 2, 65536, 8, 50)
+	case "pbkdf2":
+		fallthrough
+	default:
+		tempPasswd = pbkdf2.Key([]byte(passwd), []byte(salt), 10000, 50, sha256.New)
+	}
+
 	return fmt.Sprintf("%x", tempPasswd)
 }
 
-// HashPassword hashes a password using PBKDF.
+// HashPassword hashes a password using in PASSWORD_HASH_ALGO defined algorithm.
 func (u *User) HashPassword(passwd string) {
-	u.Passwd = hashPassword(passwd, u.Salt)
+	u.PasswdHashAlgo = setting.PasswordHashAlgo
+	u.Passwd = hashPassword(passwd, u.Salt, setting.PasswordHashAlgo)
 }
 
 // ValidatePassword checks if given password matches the one belongs to the user.
 func (u *User) ValidatePassword(passwd string) bool {
-	tempHash := hashPassword(passwd, u.Salt)
-	return subtle.ConstantTimeCompare([]byte(u.Passwd), []byte(tempHash)) == 1
+	tempHash := hashPassword(passwd, u.Salt, u.PasswdHashAlgo)
+
+	if (subtle.ConstantTimeCompare([]byte(u.Passwd), []byte(tempHash)) == 1) ||
+		(bcrypt.CompareHashAndPassword([]byte(u.Passwd), []byte(passwd)) == nil) {
+		if u.PasswdHashAlgo != setting.PasswordHashAlgo {
+			u.HashPassword(passwd)
+			UpdateUserCols(u, "passwd", "passwd_hash_algo")
```

Comment:
The method should only do validate but not update password.

Reply:
I think its useful to automatically update the password hash when the hash algo was changed without user intervention and only if the passsword is correct. Is there a better way to do this?

Reply:
I agree that it is ok to update to new hashing password if server setting has changed

Reply:
Sorry, I can't agree with you. If you want to to update database on a function nam has `validate` prefix, please rename it.

Reply:
@lunny Should i rename the function from "ValidatePassword" to "UpdatePasswordAndValidate"?

Reply:
As my personal point, I would like to keep the name `ValidatePassword` and change the logic where invoke it.